Introduce guide to testing in Go

mnsl opened the pull request:
Newcomer-friendly introduction to testing in Go.

lukechampine commented on doc/Testing.md:
```diff
@@ -0,0 +1,107 @@
+### Now run some tests
+The `make test` command runs all tests (functions starting with `Test` in `_test.go` files) for each package, setting off a panic for any test that runs longer than 5s.  For verbose output, run `make test-v` (which panics after 15s instead of 5s).  Finally, `make test-long` has verbose output, only panics when a test takes 5 minutes, and also cleans up your code using `gofmt` and `golint`.  You s ...
+
+Run `make cover` to run all tests for each package and generate color-coded .html visualizations of test coverage by function for each source file.  Run `open cover/<module>.html` to inspect a module's test coverage in your default browser.
```
`open` is OS X-specific. I don't think there's a generic POSIX command for this. Maybe just "open `cover/<module>.html` in your browser"

DavidVorick commented:
Some of the things covered here are Sia specific, for example our Makefile is pretty much only going to be found in the Sia repo.

I think I would rather see this titled something more directive, like 'How to Write Tests for Sia', given that the end goal of this document is to get people to start making pull requests, whether to add test coverage or help some other way.

DavidVorick commented:
I'm also not sure that we intend on keeping the `test-unit` and `test-integration` distinctions, they probably don't need to appear in this guide

mnsl commented:
Latest commit addresses code review comments, reorganizes layout, and adds more sections (e.g. 'A few guidelines', 'Basic test format', and 'Questions?'.

DavidVorick commented:
Need to name the file 'Running and Writing Tests for Sia.md' so that github knows render the markdown.

mnsl commented:
Fixed formatting issues (code wraps at 80 characters, added .md extension to filename).  Sorry about that.

lukechampine commented on doc/Running and Writing Tests for Sia.md:
````diff
@@ -0,0 +1,278 @@
+$ rm cover/<package>.out
+```
+Now you can open `cover/<package>.html` in a browser to view test coverage for
+<package>.  On OS X, you can just do `open cover/<package.html>`.
````
Actually, it's much simpler than that. To test a single package, run `make test pkgs=./path/to/package`. Ditto for `make bench` and `make cover`. Furthermore, you can test a single function using `make test pkgs=./package run=TestFoo`. This is made possible by the `pkgs` and `run` variables in the Makefile.

lukechampine commented on doc/Running and Writing Tests for Sia.md:
```diff
@@ -0,0 +1,278 @@
+	// Define a slice of test inputs.
+	testInputs := []int32{0, 1, -1, 42, 1234, -603}
+	// Loop over testInputs.
+	for _, in := range testInputs {
```
this is basically a table-driven test. Perhaps for better contrast with the following section, this example should be a different sort of test. Maybe something like this:

```go
func TestFoo(t *testing.T) {
    foo, err := NewFoo()
    if err != nil {
        // if NewFoo failed, we can't continue testing
        t.Fatal(err)
    }

    // try a normal input; should succeed
    err := foo.Bar(3)
    if err != nil {
        // report the error, but don't abort the test
        t.Error(err)
    }

    // try a bad input; should return an error
    // NOTE: always prefer to compare to a specific error, rather than err == nil
    err = foo.Bar(0)
    if err != errDivideByZero {
        t.Errorf("expected errDivideByZero, got %v", errDivideByZero, err)
    }
}
```

lukechampine commented on doc/Running and Writing Tests for Sia.md:
```diff
@@ -0,0 +1,278 @@
+		}
+
+        // More tests (check cases where database has already been created)
+        ...
```
I suggest using a shorter example of test-driven-development, e.g. siac/parse_test.go. There's a lot of other stuff going on in this test.

mnsl commented:
Addressed Luke's code review comments (changed testing function examples and fixed section on testing particular packages)

DavidVorick commented on doc/Running and Writing Tests for Sia.md:
```diff
@@ -0,0 +1,220 @@
+
+This guide focuses on how to write tests.  To learn about making pull requests
+to submit the code you've written, see
+[doc/Guide to Contributing to Sia.md][guide].  You should also read
```
This link is dead because your other guide is not merged yet.

DavidVorick commented on doc/Running and Writing Tests for Sia.md:
```diff
@@ -0,0 +1,220 @@
+Go's comprehensive [test package][pkg/testing] makes testing straightforward,
+particularly when you use the bundled tools included in the
+[Sia makefile][makefile], including `make test`, `make cover`, `make bench`,
+and their variants.
```
Would be nice to have a screenshot here displaying one of the testing commands.

DavidVorick commented on doc/Running and Writing Tests for Sia.md:
```diff
@@ -0,0 +1,220 @@
+
+Run `make cover` to run all tests for each package and generate color-coded
+.html visualizations of test coverage by function for each source file.  Open
+`cover/<module>.html` in a browser to inspect a module's test coverage.
```
screenshot here would probably be useful as well, showing the rendered html of some covered code.

DavidVorick commented:
Other than the things mentioned, it looks pretty good at this point

mnsl commented:
Latest changes address code review comments:
* add cover tool screenshot and sample `make test-long` output
* doc/Guide to Contributing to Sia.md has new updates; should be ready to merge that soon, which will fix the dead link

* git branching instructions have been changed to reflect changes to git configuration instructions in doc/Guide to Contributing to Sia.md

lukechampine commented on doc/Running And Writing Tests for Sia.md:
```diff
@@ -0,0 +1,256 @@
+`cover/<module>.html` in a browser to inspect a module's test coverage. For 
+example, here's part of the html file generated for the persist package: 
+
+![Screenshot](/doc/covertool.png)
```
I can't tell if this is failing to render properly because the path is wrong, or simply because the PR hasn't been merged yet.

mnsl replied:
I think it's because I messed up the relative link (see `https://help.github.com/articles/relative-links-in-readmes/`).  
It should be just `![Screenshot](covertool.png)` since the link is relative to the markdown file.

mnsl commented:
Screenshot link is fixed (changed it to raw url; the relative link wouldn't have worked until the image had been merged).

mnsl commented:
Link to screenshot doc/assets/covertool.png is fixed.
